052-prometheus-metrics-reporter.md

```diff
@@ -0,0 +1,156 @@
+1. The jmx_exporter Java agent supports complex metrics mapping rules. These rules allow renaming metrics and it can make it hard to investigate issues as metrics could have different names. When metrics are renamed, the grafana dashboards we provide can’t be reused. In addition, due to the rules, it performs badly when there’s a very large number of metrics due to a lot of topic/partitions. 
+1. Recently jmx_exporter was impacted by CVEs in its dependencies (CVE-2022-38752, CVE-2022-25857).
+
+
+## Proposal
+
+The proposal is to build a metrics reporter, `PrometheusMetricsReporter`, that implements both `org.apache.kafka.common.metrics.MetricsReporter` and `kafka.metrics.KafkaMetricsReporter`, and that directly exposes metrics via an HTTP endpoint in the Prometheus format. This will be a new project/repository under the Strimzi organization.
+
+![Proposal](images/052-proposal.png)
+
+To enable the reporter, the following Kafka configurations must be set:
+
+```
+metric.reporters=com.mickaelmaison.PrometheusMetricsReporter # For KafkaMetrics
+kafka.metrics.reporters=com.mickaelmaison.PrometheusMetricsReporter # For YammerMetrics - This is only required for brokers/controllers.
+```
+
+The reporter will expose the following configurations:
+- `prometheus.metrics.reporter.port`: The HTTP port to expose the metrics. Default: `8080`
+- `prometheus.metrics.reporter.allowlist`: A comma separated list of regex patterns to specify the metrics to collect. Default: `.*`. The patterns must match the metric names emitted by the reporter.
+
+This proposal ignores supporting HTTPS as today Strimzi does not allow configuring it with jmx_exporter. Note that jmx_exporter 0.19.0 added support for HTTPS. If needed we can add it now or later to the reporter.
+Today ZooKeeper also uses jmx_exporter to expose its metrics. ZooKeeper is going away and will be replaced by KRaft controllers which also support Kafka metrics reporters.
+
+## Affected/not affected projects
+
+This affects `strimzi-kafka-operator `. Since ZooKeeper does not support metrics reporters, we should keep the current mechanism using jmx_exporter while we support ZooKeeper deployments.
+
+The reporter will be a new project and should be usable without Strimzi for both brokers and client side applications. As it's new code, it should be first release as a preview feature before becoming the preferred mechanism once KRaft is the default.
+
+In `strimzi-kafka-operator`, we will need new examples demonstrating how to use the reporter, new grafana dashboards with metric names from the reporter.
+
+## Compatibility
```

Comment:
I think this should make it clear that - at least for some time - the JMX Exporter will remain supported even for Kafka to ensure backward compatibility and eventually suggest some timeline for removal (Maybe together with ZooKeeper support?).

Reply:
Yes, since ZooKeeper does not support reporters, for at least as long as it's supported we'll need to support the JMX exporter.